python/lsst/daf/butler/registry/versions.py

```diff
@@ -0,0 +1,321 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = [
+    "ButlerVersionsManager", "IncompatibleVersionError", "MissingVersionError"
+]
+
+import hashlib
+import logging
+from typing import (
+    TYPE_CHECKING,
+    Iterable,
+    List,
+    Mapping,
+    MutableMapping,
+    Optional,
+)
+
+import sqlalchemy
+
+if TYPE_CHECKING:
+    from .interfaces import (
+        ButlerAttributeManager,
+    )
+    from ..core import Config
+
+
+_LOG = logging.getLogger(__name__)
+
+
+class MissingVersionError(RuntimeError):
+    """Exception raised when existing database is missing attributes with
+    version numbers.
+    """
+    pass
+
+
+class IncompatibleVersionError(RuntimeError):
+    """Exception raised when configured version number is not compatible with
+    database version.
+    """
+    pass
+
+
+class VersionTuple:
```

Comment:
Unless this needs to be mutable, it seems like it might make sense to actually use `typing.NamedTuple` (as the base class here).  That will get you `__init__` and what I think are probably the right comparison operators for free, and if you call the class "Tuple" it's reasonable for users to expect indexing, iteration, etc. to work (and that makes sense, given the analogy with e.g. `sys.version_info`).

Reply:
👍 I thought about just doing `collections.namedtuple` as I did not really like much of `typing` stuff, but I can do it too. This was supposed to be internal class but it looks like it's going to be more exposed (or moved somewhere) so I should make it right.

Reply:
I can understand not liking a lot of `typing`, but IMO `typing.NamedTuple` has much nicer than the original `collections.namedtuple`.  I hope you'll agree.